File: android/src/main/java/com/rnmqtt/models/PublishOptions.kt
package com.rnmqtt.models

import com.facebook.react.bridge.ReadableMap
import com.rnmqtt.utils.getOr

data class PublishOptions(
  val qos: QoS,
  val retain: Boolean,
  val isDuplicate: Boolean,
) {
  constructor(pubOptionsFromJs: ReadableMap): this(
    QoS.values()[pubOptionsFromJs.getOr<Int>("qos", 0)],
    pubOptionsFromJs.getOr("retain", false),
    pubOptionsFromJs.getOr("isDuplicate", false)
  )
}

File: ios/models/PublishOptions.swift
struct PublishOptions {
  let qos: QoS
  let retain: Bool
  let isDuplicate: Bool

  init(fromJsPubOptions pubOptionsFromJs: NSDictionary) {
    self.qos = QoS(rawValue: willFromJs["qos"] ?? 0)
    self.retain = willFromJs["retain"] ?? false
    self.isDuplicate = willFromJs["isDuplicate"] ?? false
  }
}